src/core/fut.ml

```diff
@@ -1,118 +1,114 @@
 
 let[@inline] is_success self =
-  match A.get self.st with
-  | Done (Ok _) -> true
-  | _ -> false
+  match C.peek_exn self.st with
+  | _ -> true
+  | exception _ -> false
 
-let[@inline] is_failed self =
-  match A.get self.st with
-  | Done (Error _) -> true
-  | _ -> false
+let[@inline] is_failed self = C.is_canceled self.st
 
 exception Not_ready
 
 let[@inline] get_or_fail self =
-  match A.get self.st with
-  | Done x -> x
-  | Waiting _ -> raise Not_ready
+  match C.peek self.st with
+  | Some x -> x
+  | None -> raise Not_ready
 
 let[@inline] get_or_fail_exn self =
-  match A.get self.st with
-  | Done (Ok x) -> x
-  | Done (Error (exn, bt)) -> Printexc.raise_with_backtrace exn bt
-  | Waiting _ -> raise Not_ready
+  match C.peek_exn self.st with
+  | x -> x
+  | exception C.Running -> raise Not_ready
+
+let[@inline] peek_or_assert_ (self : 'a t) : 'a =
+  match C.peek_exn self.st with
+  | x -> x
+  | exception C.Running -> assert false
+
+let on_result_cb_ _tr f self : unit =
+  match peek_or_assert_ self with
+  | x -> f (Ok x)
+  | exception exn ->
+    let ebt = Exn_bt.get exn in
+    f (Error ebt)
 
 let on_result (self : _ t) (f : _ waiter) : unit =
-  while
-    let st = A.get self.st in
-    match st with
-    | Done x ->
-      f x;
-      false
-    | Waiting { waiters = l } ->
-      not (A.compare_and_set self.st st (Waiting { waiters = f :: l }))
-  do
-    Domain_.relax ()
-  done
+  let trigger =
+    (Trigger.from_action f self on_result_cb_ [@alert "-handler"])
+  in
+  if not (C.try_attach self.st trigger) then on_result_cb_ () f self
 
-exception Already_fulfilled
+let on_result_ignore_cb_ _tr f (self : _ t) =
+  f (Picos.Computation.canceled self.st)
 
-let fulfill (self : _ t) (r : _ result) : unit =
-  let fs = ref [] in
-  while
-    let st = A.get self.st in
-    match st with
-    | Done _ -> raise Already_fulfilled
-    | Waiting { waiters = l } ->
-      let did_swap = A.compare_and_set self.st st (Done r) in
-      if did_swap then (
-        (* success, now call all the waiters *)
-        fs := l;
-        false
-      ) else
-        true
-  do
-    Domain_.relax ()
-  done;
-  List.iter (fun f -> try f r with _ -> ()) !fs;
-  ()
+let on_result_ignore (self : _ t) f : unit =
+  if Picos.Computation.is_running self.st then (
+    let trigger =
+      (Trigger.from_action f self on_result_ignore_cb_ [@alert "-handler"])
+    in
+    if not (C.try_attach self.st trigger) then on_result_ignore_cb_ () f self
+  ) else
+    on_result_ignore_cb_ () f self
 
 let[@inline] fulfill_idempotent self r =
-  try fulfill self r with Already_fulfilled -> ()
+  match r with
+  | Ok x -> C.return self.st x
+  | Error ebt -> C.cancel self.st ebt
+
+exception Already_fulfilled
+
+let fulfill (self : _ t) (r : _ result) : unit =
+  let ok =
+    match r with
+    | Ok x -> C.try_return self.st x
+    | Error ebt -> C.try_cancel self.st ebt
+  in
+  if not ok then raise Already_fulfilled
```

Comment:
It's not obvious that this is the only reason for a `not ok`, e.g. when `try_cancel` fails.

Reply:
`try_cancel` should only fail if it's already fulfilled, I think?